File: .semgrep/tests/immutability/python/no_mutation_test.py
# ruleid: immutability.no-list-append-param-python
def add_item(items, item):
    items.append(item)
    return items

# ruleid: immutability.no-list-extend-param-python
def extend_items(items, other):
    items.extend(other)
    return items

# ruleid: immutability.no-dict-mutation-param-python
def update_user(user, name):
    user["name"] = name
    return user

# ruleid: immutability.no-list-sort-param-python
def sort_items(items):
    items.sort()
    return items

# ok: immutability.no-list-append-param-python
def add_item_safe(items, item):
    return [*items, item]

# ok: immutability.no-dict-mutation-param-python
def update_user_safe(user, name):
    return {**user, "name": name}

# ok: immutability.no-list-sort-param-python
def sort_items_safe(items):
    return sorted(items)

File: .semgrep/tests/immutability/go/no_mutation_test.go
package main

// ruleid: immutability.no-param-slice-append-go
func AddItem(items []string, item string) []string {
	return append(items, item)
}

// ruleid: immutability.no-pointer-field-mutation-go
func UpdateUser(user *User, name string) *User {
	user.Name = name
	return user
}

// ok: immutability.no-param-slice-append-go
func AddItemSafe(items []string, item string) []string {
	newItems := make([]string, len(items)+1)
	copy(newItems, items)
	newItems[len(items)] = item
	return newItems
}

// ok: immutability.no-pointer-field-mutation-go
func UpdateUserSafe(user User, name string) User {
	return User{
		Name: name,
		Age:  user.Age,
	}
}

type User struct {
	Name string
	Age  int
}
